tests/test_halo_data_transformer.py

```diff
@@ -0,0 +1,471 @@
+            quantity.origin,
+            quantity.extent,
+            quantity.data.shape,
+            direction,
+            n_halos,
+            interior=False,
+        )
+
+    return send_boundaries, recv_boundaries
+
+
+def test_data_transformer_scalar_pack_unpack(quantity, rotation, n_halos):
+    original_quantity: Quantity = copy.deepcopy(quantity)
+
+    send_boundaries, recv_boundaries = _get_boundaries(quantity, n_halos)
+    N_edge_boundaries = {
+        0: (send_boundaries[NORTH], recv_boundaries[SOUTH]),
+        -1: (send_boundaries[NORTH], recv_boundaries[WEST]),
+        -2: (send_boundaries[NORTH], recv_boundaries[NORTH]),
+        -3: (send_boundaries[NORTH], recv_boundaries[EAST]),
+    }
+
+    NE_corner_boundaries = {
+        0: (send_boundaries[NORTHEAST], recv_boundaries[SOUTHEAST]),
+        -1: (send_boundaries[NORTHEAST], recv_boundaries[SOUTHWEST]),
+        -2: (send_boundaries[NORTHEAST], recv_boundaries[NORTHWEST]),
+        -3: (send_boundaries[NORTHEAST], recv_boundaries[NORTHEAST]),
+    }
+
+    specification = HaloUpdateSpec(
+        n_halo_points=n_halos,
+        shape=quantity.data.shape,
+        strides=quantity.data.strides,
+        itemsize=quantity.data.itemsize,
+        origin=quantity.metadata.origin,
+        extent=quantity.metadata.extent,
+        dims=quantity.metadata.dims,
+        numpy_module=quantity.np,
+        dtype=quantity.metadata.dtype,
+    )
+
+    exchange_descriptors = [
+        HaloExchangeData(
+            specification,
+            N_edge_boundaries[rotation][0],
+            rotation,
+            N_edge_boundaries[rotation][1],
+        ),
+        HaloExchangeData(
+            specification,
+            NE_corner_boundaries[rotation][0],
+            rotation,
+            NE_corner_boundaries[rotation][1],
+        ),
+    ]
+
+    data_transformer = HaloDataTransformer.get(quantity.np, exchange_descriptors)
+
+    data_transformer.async_pack([quantity, quantity])
+    data_transformer.synchronize()
+    # Simulate data transfer
+    data_transformer.get_unpack_buffer().assign_from(
+        data_transformer.get_pack_buffer().array
+    )
+    data_transformer.async_unpack([quantity, quantity])
+    data_transformer.finalize()
+
+    # From the copy of the original quantity we rotate data
+    # according to the rotation & slice and insert them bak
+    # this reproduce the multi-buffer strategy
+    rotated = rotate_scalar_data(
+        original_quantity.data[N_edge_boundaries[rotation][0]],
+        original_quantity.dims,
+        original_quantity.metadata.np,
+        -rotation,
+    )
+    original_quantity.data[N_edge_boundaries[rotation][1]] = rotated
+    rotated = rotate_scalar_data(
+        original_quantity.data[NE_corner_boundaries[rotation][0]],
+        original_quantity.dims,
+        original_quantity.metadata.np,
+        -rotation,
+    )
+    original_quantity.data[NE_corner_boundaries[rotation][1]] = rotated
```

Comment:
I try to reserve "original_<thing>" in tests for the unchanged input to make this clear when doing comparisons at the end, but here the original gets modified before comparing to the output. Could you instead name the comparison quantity you create with deepcopy "target_quantity"? You can give `quantity` as input to these rotate routines, since they don't mutate the input.

Reply:
Done